lib/virtual-fs/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+    fn set_last_modified(&self, _last_modified: u64) {
+        debug!("{:?} did nothing in VirtualFile::set_last_modified due to using the default implementation", self);
+    }
+
+    /// set the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn set_created_time(&self, _created_time: u64) {
+        debug!(
+            "{:?} did nothing in VirtualFile::set_created_time to using the default implementation",
+            self
+        );
+    }
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError>;
+
+    /// Request deletion of the file
+    fn unlink(&mut self) -> Result<(), FsError>;
+
+    /// Store file contents and metadata to disk
+    /// Default implementation returns `Ok(())`.  You should implement this method if you care
+    /// about flushing your cache to permanent storage
+    fn sync_to_disk(&self) -> Result<(), FsError> {
+        Ok(())
+    }
+
+    /// Moves the file to a new location
+    /// NOTE: the signature of this function will change before stabilization
+    // TODO: stablizie this in 0.7.0 or 0.8.0 by removing default impl
+    fn rename_file(&self, _new_name: &std::path::Path) -> Result<(), FsError> {
+        panic!("Default implementation for now as this method is unstable; this default implementation or this entire method may be removed in a future release.");
+    }
+
+    /// Returns the number of bytes available.  This function must not block
+    fn bytes_available(&self) -> Result<usize, FsError>;
+
+    /// Used for polling.  Default returns `None` because this method cannot be implemented for most types
+    /// Returns the underlying host fd
+    fn get_raw_fd(&self) -> Option<i32> {
+        None
+    }
+}
+
+// Implementation of `Upcastable` taken from https://users.rust-lang.org/t/why-does-downcasting-not-work-for-subtraits/33286/7 .
+/// Trait needed to get downcasting from `VirtualFile` to work.
+pub trait Upcastable {
+    fn upcast_any_ref(&'_ self) -> &'_ dyn Any;
+    fn upcast_any_mut(&'_ mut self) -> &'_ mut dyn Any;
+    fn upcast_any_box(self: Box<Self>) -> Box<dyn Any>;
+}
+
+impl<T: Any + fmt::Debug + 'static> Upcastable for T {
+    #[inline]
+    fn upcast_any_ref(&'_ self) -> &'_ dyn Any {
+        self
+    }
+    #[inline]
+    fn upcast_any_mut(&'_ mut self) -> &'_ mut dyn Any {
+        self
+    }
+    #[inline]
+    fn upcast_any_box(self: Box<Self>) -> Box<dyn Any> {
+        self
+    }
+}
+
+impl dyn VirtualFile + 'static {
+    #[inline]
+    pub fn downcast_ref<T: 'static>(&'_ self) -> Option<&'_ T> {
+        self.upcast_any_ref().downcast_ref::<T>()
+    }
+    #[inline]
+    pub fn downcast_mut<T: 'static>(&'_ mut self) -> Option<&'_ mut T> {
+        self.upcast_any_mut().downcast_mut::<T>()
+    }
```

Comment:
I believe this can simplified a bit if we have a `as_any` method? (and we move the downcasting back to the code that uses and needs it)

Reply:
Well, we're dealing with `Box<dyn VirtualFile>` (previously `Box<dyn WasiFile>`), due to the way this trait works, everything just works and users don't have to do anything but they can use these methods if they want, so I think it's pretty good. We can discuss possible improvements though